src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,272 @@
+            {
+                if (!_makeOuterTypesPartial)
+                {
+                    return null;
+                }
+
+                // we recognize filenames of that follow the pattern "OuterClass.InnerClass.cs" 
+                // and generate code for this file, which will look like so:
+                // a partial definition of OuterClass and a nested private InnerClass.
+                var nameParts = Path.GetFileNameWithoutExtension(newDocumentName).Split(new char[] { '.' });
+                if (nameParts == null || nameParts.Count() != 2)
+                {
+                    return null;
+                }
+
+                var innerTypeName = nameParts[1];
+                return generator.ClassDeclaration(innerTypeName, accessibility: Accessibility.Private);
+            }
+
+            // TODO: look to optimize the number of Ancestors and Descendants calls here.
+            private static IEnumerable<SyntaxNode> GetMembersToRemove(
+                SyntaxNode root, TTypeDeclarationSyntax typeNode, bool removeDescendentsOfSelf)
+            {
+                var ancestorsAndSelfToKeep = typeNode
+                    .AncestorsAndSelf()
+                    .Where(n => n is TNamespaceDeclarationSyntax || n is TTypeDeclarationSyntax);
+
+                var topLevelMembersToRemove = root
+                    .DescendantNodesAndSelf(descendIntoChildren: _ => true, descendIntoTrivia: false)
+                    .Where(n => IsTopLevelNamespaceOrTypeNode(n)
+                                ? !ancestorsAndSelfToKeep.Contains(n)
+                                : false);
+
+                var ancestorsKept = ancestorsAndSelfToKeep
+                    .OfType<TTypeDeclarationSyntax>()
+                    .Except(new[] { typeNode });
+
+                var membersOfAncestorsKept = ancestorsKept
+                    .SelectMany(a => a.DescendantNodes().OfType<TMemberDeclarationSyntax>().Where(t => !t.Equals(typeNode) && t.Parent.Equals(a)));
+
+                var membersToRemove = topLevelMembersToRemove.Concat(membersOfAncestorsKept);
+
+                if (removeDescendentsOfSelf)
+                {
+                    var descendentsOfSelf = typeNode
+                        .DescendantNodes(descendIntoChildren: _ => true, descendIntoTrivia: false)
+                        .OfType<TMemberDeclarationSyntax>();
+
+                    membersToRemove = membersToRemove.Concat(descendentsOfSelf);
+                }
+
+                return membersToRemove;
+            }
+
+            private static bool IsTopLevelNamespaceOrTypeNode(SyntaxNode node)
+            {
+                return ((node is TNamespaceDeclarationSyntax && node.Parent is TCompilationUnitSyntax)
+                     || (node is TTypeDeclarationSyntax && (node.Parent is TNamespaceDeclarationSyntax
+                                                            || node.Parent is TCompilationUnitSyntax)));
```

Comment:
"(node.Parent is CompilationUnit) || (node is Type and node.Parent is Namespace)"